### PR TITLE
[WIP] Extract prediction into separate interface

### PR DESCRIPTION
**Pull request by RAMitchell**
Remove prediction logic from gbtree class and extract into new prediction interface.

This will allow us to perform prediction on GPUs in future. 

This pr should contain no changes to the execution of XGBoost.

### CODE REVIEW

**Comment by khotilov** [comment]
@RAMitchell could you please explain your motivations for so many stylistic changes in one go, and not even in a separate commit? I see very little need for lots of them, and some are even detrimental (mangling of doxygen comments). And it makes merging a royal pain for other contributors working on the same code.

**Comment by RAMitchell** [comment]
Sorry - changes are from clang format, I will reapply changes over the original formatting.

**Comment by RAMitchell** [comment]
I have reverted the formatting changes.

**Comment by tqchen** [comment]
This seems to be a bit over complicated for a predictor. For CPU prediction, a single predictor interface might be useful. Also we could hide the tree structure into the interface. Something like Learner.GetPredictor(), so the interface does not have to contain model.


cc @hcho3 who is doing something relevant

**Comment by hcho3** [comment]
@RAMitchell @tqchen I agree that it is best to separate prediction logic from GBTree. For now, I think we can add a simple predictor interface to xgboost.

On the longer term, it would be best to create a new project that's dedicated to fast, efficient prediction. I've been working on a new project provisionally called **tree-lite**, which will be made public in the near future.

Some facts about tree-lite:
* It is an independent project, to be hosted under the dmlc group.
* It will support a **uniform predictor interface** for multiple tree formats, such as those produced by LightGBM and xgboost.
* Many optimization options will be offered, such as if-else rule compilation.

**Comment by RAMitchell** [comment]
@tqchen 
I want to clarify that I was not intending to expose a predictor class or interface in the C API itself (e.g. to have a predictor class in python). The purpose of this PR was to extract the prediction logic from the gbtree class (which is somewhat monolithic).

>This seems to be a bit over complicated for a predictor.
When you say overcomplicated do you refer to the fact that there are 4 different functions or that there are so many function parameters?

The first three methods in my predictor interface could be combined into something more like the learner interface:
```
void Predict(DMatrix* data, bool output_margin,
               std::vector<bst_float>* out_preds, unsigned ntree_limit,
               bool pred_leaf, bool pred_contribs) const override {
```

I don't like using multiple booleans in a function like this because it is unclear what the precedence of each argument is. Maybe an enum is a better solution to select the type of prediction?

>Also we could hide the tree structure into the interface. Something like Learner.GetPredictor(), so the interface does not have to contain model.

Can you please explain this more? 

I would definitely like to simplify the interface. Conceptually a prediction function should only need three arguments - the input data, a model and the output predictions. The problem as I see it is that we don't really have an independent 'model' class, GBTree contains the model but it also contains a lot of other stuff. I think encapsulation would be easier if this model class existed. The implementation of dropout (DART) would also become much simpler.

@hcho3 
 Excited to hear about this. I could potentially be interested in contributing GPU prediction.

Does tree-lite solve the problem of test set evaluation during training, for example in cross validation? I know that compiling code on the fly can sometimes be a net performance loss unless the compiled code needs to be executed very many times.

**Comment by superbobry** [comment]
@hcho3 sorry for hijacking the discussion, but I'm curious if you have considered implementing [QuickScorer](http://dl.acm.org/citation.cfm?id=2767733) as an alternative to if-else/compilation-based approaches?

**Comment by hcho3** [comment]
@superbobry Wow this is very interesting! Thanks a lot for the link. In general, if-else compilation is only one of many possible ways to tackle fast prediction problem. My goal is to provide a polished package that lets practitioners experiment rapidly with different ways to optimize tree prediction.

@ramitchell For now I'm focusing on the deployment scenario where it makes sense to compile a (finished) model into optimized prediction. But I do acknowledge there's an inherent trade-off between interpreting vs compiling rules. I can see providing an interpretation-only option when compilation time is problematic.

**Comment by superbobry** [comment]
@hcho3 glad this is useful. Look forward to seeing the library you're working on open-sourced. I'd be happy to contribute.

**Comment by tqchen** [comment]
@RAMitchell I think the main problem I have with this is that too many arguments being passed to the predictor, while the predictor itself seems to be a function, rather than a stateful object, kills the need of the class(except for the GPU logic).

Another question I have is how would this handle the old caching behavior.  If the major reason is to enable GPU related logic, I think I agree we could have something minimized, and legitimate predictor interface like updater.

**Comment by RAMitchell** [comment]
Some changes in the latest commit:
- Moved prediction cache into the predictor class
- Created a GBTreeModel struct which encapsulates related member variables from GBTree and also implements related functions (model load/save)
- Reduced number of parameters to predictor functions by passing GBTreeModel

The end result is a dramatically smaller GBTree class that now contains mostly boosting logic.

We still have some code duplication with the DART booster. My preferred solution for this is to merge the DART booster completely into GBTree. I think this would only need minor changes in functions and would have no performance impact. The default parameters for DART are equivalent to ordinary gradient boosting anyway.

The only issue with this is the serialisation of the model. If we change this it may not be possible to load models saved from the previous XGBoost version.

**Comment by hcho3** [comment]
Sorry in advance for hijacking the thread.
@superbobry I just found out that the QuickScorer algorithm has been patented. (See [this](https://github.com/hpclab/quickscorer) and [this](patentscope.wipo.int/search/en/WO2016203501).) It certainly makes it difficult to build bitvector traveral algorithm into tree-lite in the near future. We'd need legal counsel at least before proceeding further.

**Comment by tqchen** [comment]
We do need backward compatibility

**Comment by RAMitchell** [comment]
Ok I will leave the dart stuff for now. Any comments about the other changes? I'd like to merge and start work on gpu prediction.

**Comment by tqchen** [comment]
Do a sweep over the predictor interface docstrings, some of them are missing. Also state clearly what state should be kept in predictor(cache), vs what should be kept in model(GBM) in the comments of predictor interface

File: include/xgboost/predictor.h [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+ * \brief Interface of predictor,
+ *  performs predictions for a gradient booster.
+ */
+#pragma once
+#include <xgboost/base.h>
+#include <functional>
+#include <memory>
+#include <vector>
+#include <string>
+#include "../../src/gbm/gbtree_model.h"
```
Comment (khotilov) [inline include/xgboost/predictor.h]: Should we move gbtree_model.h to include/xgboost/ ?

File: include/xgboost/predictor.h [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+ *        Prediction functions all take a GBTreeModel and a DMatrix as input and output a vector of
+ *        predictions. The predictor does not modify any state of the model itself.
+ */
+
+class Predictor {
+ public:
+  virtual ~Predictor() {}
+
+  /**
+   * \fn  void Predictor::InitCache(const std::vector<std::shared_ptr<DMatrix> > &cache);
```
Comment (khotilov) [inline include/xgboost/predictor.h]: \fn is uncommon and is rarely needed 
https://www.star.bnl.gov/public/comp/sofi/doxygen/commands.html#cmdfn
Was there any specific reason for using it in this code?

Comment (RAMitchell) [inline include/xgboost/predictor.h]: This was automatically generated. I will avoid doing this in future in that case.
